hashintel/hash
# Add block config panel. Add block schema to block context

## 🌟 What is the purpose of this PR?

Add a configuration panel which provides a generic UI for setting block properties outside of the block. At the moment, this is only for properties listed in a special `configProperties` key in the block's schema.

This will be useful for blocks which don't expose their own configuration UI (e.g. existing components which have been wrapped), or where they could but ideally wouldn't (because it introduces clutter), and likely other things.

This PR also fixes draft id creation and renames and tweaks functions related to draftIds and entities for clarity.

**Because this updates how blocks are cached, you must remove `hash-workspace-user-blocks` from local storage before testing, and after this is merged.**

## ⚠️ Known issues
- `BlockView` now takes its block data from the draft entity store, so that it subscribes to the latest user changes. There are currently two issues with this:
  - the call to `updateEntityProperties` from the config menu causes an update to the draft entity store, but does not actually cause this change to be persisted to the API and therefore to the saved entity store. This is because the `save` function in collab is not checking for changes in entities whose content are not in the PM tree. This is being fixed separately (tracked in [this task](https://app.asana.com/0/1200211978612931/1202106892392942/f) _(internal)_)
  - if a configuration property is edited from within a block, rather than the menu, there is a short delay between the edit being made and the updated property value being reflected in the configuration menu. This is a consequence of updates from some blocks not going via collab, which won't be fixed in this PR, but will be when we send all updates via collab.

## 🐾 Next steps

- Add `configProperties` to the Block Protocol types and spec.
- The config panel is not where it will likely end up, and definitely not how it will look. It should end up re-using the same inputs as the entity editor, as it's essentially an entity editor for a subset of the entity's properties.
- Fixes to properly cover all entities in collab, as mentioned above

## 🔍 What does this change?
- Adds a `BlockConfigMenu` triggered via the context menu, which shows inputs for any `configProperties` listed in the block's schema
- Updates `UserBlocksContext` to include the block schema as well as metadata, using `blocksMeta`, and gets rid of the duplicative `RemoteBlockMetadata` type that was previously used.
- Fixes the code block to send an update when the language is changed (it previously only updated local state)
- Get rid of `draftIdForEntity`, which was broken since we introduced fake draft ids in #480.
- Renames and updates functions 'getting a draft id by entity id' for clarity and consistency.
- Adds some comments where I thought they'd be helpful for future travellers

DX improvements for testing blocks:
- overrides cached blocks with new versions of the same blocks in all situations
- allows for reloading a block from URL (rather than preventing this once it is has been defined)

## 🔗 Related links





- [Asana task for config panel](https://app.asana.com/0/1201959586244685/1202107700561474/f) _(internal)_
- [Asana task for merging blocksMeta and initialUserBlocks](https://app.asana.com/0/1200211978612931/1202023490862451/f) _(internal)_

## ❓ How to test this?

1.  Pick a block, e.g. `code`, and run `yarn build && yarn serve` in its folder
2. Edit `dist/block-schema.json` to add 
```json
    "configProperties": [ "content", "language", "caption" ]
```
3. Paste `http://localhost:62681` into the 'load block from URL'
4. Change some property, wait a sec and see it update.

## 📹 Demo

https://user-images.githubusercontent.com/37743469/163582249-a169af1e-3941-40e9-a234-e363e6dfe521.mp4
